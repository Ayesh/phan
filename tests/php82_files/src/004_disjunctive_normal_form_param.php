<?php
namespace NS4;
class A {
    public function example() {
        echo "done\n";
    }
}
class B {
}
interface C {}
class D extends B implements C {}
function example(A|(B&C) $param): A|(B&C) {
    '@phan-debug-var $param';
    $param->example('extra');
    if (random_int(0,1)) {
        return new B();
    }
    return $param;
}
example(new A()); // valid
example(new B()); // invalid
example(new D()); // valid
example(new \stdClass()); // invalid
example(null); // invalid
