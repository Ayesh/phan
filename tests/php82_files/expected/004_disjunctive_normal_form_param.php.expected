%s:4 PhanPluginUnknownMethodReturnType Method \NS4\A::example() has no declared or inferred return type
%s:12 PhanDebugAnnotation @phan-debug-var requested for variable $param - it has union type (\NS4\B&\NS4\C)|\NS4\A(real=(\NS4\B&\NS4\C)|\NS4\A)
%s:14 PhanParamTooMany Call with 1 arg(s) to \NS4\A::example() which only takes 0 arg(s) defined at %s:4
%s:16 PhanTypeMismatchReturnSuperType Returning new B() of type \NS4\B but example() is declared to return (\NS4\B&\NS4\C)|\NS4\A (saw a supertype instead of a subtype)
%s:21 PhanTypeMismatchArgumentSuperType Argument 1 ($param) is new B() of type \NS4\B but \NS4\example() takes (\NS4\B&\NS4\C)|\NS4\A defined at %s:12 (expected type to be the same or a subtype, but saw a supertype instead)
%s:23 PhanTypeMismatchArgument Argument 1 ($param) is new \stdClass() of type \stdClass but \NS4\example() takes (\NS4\B&\NS4\C)|\NS4\A defined at %s:12
%s:24 PhanTypeMismatchArgumentReal Argument 1 ($param) is null of type null but \NS4\example() takes (\NS4\B&\NS4\C)|\NS4\A defined at %s:12
